### PR TITLE
Fix: Update Android firebase/GA4 to support Google EU Consent

### PR DESCRIPTION
**Pull request by Mansi-mParticle**
## Instructions
 1. PR target branch should be against `development`
 2. PR title name should follow this format: https://github.com/mParticle/mparticle-workflows/blob/main/.github/workflows/pr-title-check.yml
 3. PR branch prefix should follow this format: https://github.com/mParticle/mparticle-workflows/blob/main/.github/workflows/pr-branch-check-name.yml

 ## Summary
 - Implemented Google EU Consent for GA4

 ## Testing Plan
 - [ ] Was this tested locally? If not, explain why.
 - Verified through compilation and execution of a local application, as well as running the unit test case to validate the functionality.

 ## Reference Issue (For mParticle employees only.  Ignore if you are an outside contributor)
 - Closes https://go.mparticle.com/work/SQDSDKS-6176

### CODE REVIEW

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (alexs-mparticle) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:903]: Maybe I'm missing something, but should this function be empty?

File: src/main/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4Kit.kt
Comment (samdozor) [inline src/main/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4Kit.kt:36]: if you wanted to be extremely cautious - you should grab a reference and check if the reference is null. otherwise in your code it's possible that the `currentuser` or the `consentstate` becomes null (due to another thread modifying it) in between lines. not sure if this compile but here's the gist of how i would do it:
```suggestion
        val user = currentUser?.consentState?
        user?.let {
            setConsent(user)
        }
```

File: src/main/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4Kit.kt
Comment (samdozor) [inline src/main/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4Kit.kt:553]: add test coverage for this please

File: src/main/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4Kit.kt
Comment (samdozor) [inline src/main/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4Kit.kt:570]: add test coverage for this as well

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (samdozor) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:903]: empty test

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (alexs-mparticle) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:176]: Pardon my ignorance since I'm not as fluent in Kotlin, but this looks to me like the test isn't really testing an actual updated consent state.

If I'm reading this correctly, the test case starts with `defaultAdPersonalizationConsentSDK` being set to `Denied` via the server configuration, and then `kitInstance.onConsentStateUpdated` us called with `locationCollectionConsent` being set to `false`.

If `kitInstance.onConsentStateUpdated(state, state, filteredMParticleUser)` is never called, will the assertion still be accurate?

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (alexs-mparticle) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:198]: I don't think CCPA consent is relevant as a test as it does not pertain to the EU consent fix.

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (alexs-mparticle) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:214]: Like my previous question about the test cases, this test function is called `onConsentStateUpdatedTest_When_Marketing_true`, so I would expect that this assertion should be `TestCase.assertEquals("GRANTED", expectedConsentValue)`, but I don't see step in this test that sets `Marketing` to `true` or `granted`.

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (alexs-mparticle) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:236]: Again, I don't think you need to include CCPA unless you are testing that adding extra consent states doesn't affect the outcome. However that does not appear to be the case.

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (alexs-mparticle) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:325]: This might be a nitpick on my part, but I would rename `locationCollectionConsent` and `parentalConsent` to `marketingConsent` and `performanceConsent` for readability.

**Review by samdozor — CHANGES_REQUESTED** [review]
as we'e discussed - also need to add the firebase dependency to the test target/module

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (alexs-mparticle) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:457]: It looks like you're only testing the actual value of `AD_STORAGE` but not the other three default settings, but the next assertion seems to test if all 4 are in the payload. Can you write assertions for all 4 values individually?

**Review by alexs-mparticle — CHANGES_REQUESTED** [review]
@Mansi-mParticle I have some questions and feedback. I'm not too familiar with some of the Kotlin conventions, so I apologize if some of my questions seem pretty basic.

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (alexs-mparticle) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:547]: I noticed that you have tests here for cases where the JSON is invalid or empty, but I don't see a test for the "happy path" where it correctly parses JSON. Is that not necessary?

**Review by alexs-mparticle — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4Kit.kt
Comment (einsteinx2) [inline src/main/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4Kit.kt:34]: Should this be called `val consentState` instead of `val user`? What is the type of `user` here?

File: src/main/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4Kit.kt
Comment (einsteinx2) [inline src/main/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4Kit.kt:856]: The parameter here is called `values` but it seems it's only a single value. Could this be renamed to `value` to be more clear? With `values` it makes me expect an array or something.

File: src/test/kotlin/com/google/firebase/analytics/FirebaseAnalytics.kt
Comment (einsteinx2) [inline src/test/kotlin/com/google/firebase/analytics/FirebaseAnalytics.kt:28]:
```suggestion

    fun setConsent(var1:  MutableMap<Any, Any>) {
```
Nitpick about spacing.

File: src/test/kotlin/com/google/firebase/analytics/FirebaseAnalytics.kt
Comment (einsteinx2) [inline src/test/kotlin/com/google/firebase/analytics/FirebaseAnalytics.kt:14]: Nitpick about spacing. Looks like these two added lines now make 3 lines of whitespace, maybe remove lines 13 and 14?

I tried to do it with a change suggestion but github wasn't parsing it correctly.

File: src/test/kotlin/com/google/firebase/analytics/FirebaseAnalytics.kt
Comment (einsteinx2) [inline src/test/kotlin/com/google/firebase/analytics/FirebaseAnalytics.kt:36]:
```suggestion
    }
    
```

More spacing nitpicks :P

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (einsteinx2) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:43]: These two lines as well (43 and 44)

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (einsteinx2) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:55]: And this line 55

File: src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt
Comment (einsteinx2) [inline src/test/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseGA4KitTest.kt:142]: And this line 142

**Review by einsteinx2 — CHANGES_REQUESTED** [review]
I'll defer to Alex on GA4 consent behavior, so mostly focusing on code clarity and formatting. Apologies for all the spacing nitpicks but it's a pet peeve of mine :)

Great to see extensive unit testing!

**Review by BrandonStalnaker — APPROVED** [review]
Unit Testing looks great! The logic seems all correct to me

**Review by alexs-mparticle — APPROVED** [review]
Looks good to me. Once @einsteinx2 and @samdozor sign off on their comments I think this is good to merge.

**Review by einsteinx2 — APPROVED** [review]

**Review by samdozor — CHANGES_REQUESTED** [review]
Code/functionality looks good! Just some feedback on error messages so help our customers and support team better debug when things go wrong.

**Review by samdozor — APPROVED** [review]

**Pull request merged**
